### PR TITLE
blog: add blog introducing the deadlock gadget (LFX Mentorship)

### PR DESCRIPTION
**Pull request by Snehil-Shah**
Adds blog announcing the deadlock gadget as part of LFX Mentorship

cc: @alban @burak-ok

### CODE REVIEW

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+![A diagram depicting a deadlock](/media/2024-12-06-deadlock-depiction.svg)
+
+Thread 1, currently using mutex 1, requests access to lock mutex 2, but has to wait as mutex 2 is being used by thread 2. Now thread 2 requests access to mutex 3, but also has to wait as mutex 3 is being used by thread 3. Finally, thread 3 requests access to mutex 1, but has to wait as, in case you forgot, mutex 1 is being used by thread 1.
+We have come full circle ending up with the threads being stuck in an infinite wait cycle due to mutex lock order inversions, thus hanging the process.
+
+:::tip
+
+The diagram above represents a directed graph. Keep reading to learn how we use it to detect these deadlocks.
+
+:::
```
Comment (Snehil-Shah) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:44]: This syntax is with the presumption that we are using docusaurus ([ref](https://docusaurus.io/docs/next/markdown-features/admonitions)).

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+authors: [snehil-shah]
+description: "Introducing a new gadget for detecting deadlocks"
+draft: false
+tags: ["eBPF", "ig", "Inspektor Gadget", "Gadget", "WASM", "deadlock", "LFX Mentorship"]
+title: "Introducing a new gadget for detecting deadlocks"
+slug: /2024/12/deadlock-gadget-lfx-mentorship
+image: /media/2024-05-31-cncf-mentoring-logo.jpg
+---
+
+Hey! I am [Snehil Shah](https://github.com/Snehil-Shah), a newer inspektor in the community. Over these past few months, I have been working under the mentorship of [Alban Crequy](https://github.com/alban) and [Burak Ok](https://github.com/burak-ok) as part of the [LFX Mentorship](https://mentorship.lfx.linuxfoundation.org/project/7fdda09c-0eb8-466b-9fdf-e4b3c6a1d5b3) program.
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:11]:
```suggestion
Hey! I am [Snehil Shah](https://github.com/Snehil-Shah), a new inspektor in the community. Over these past few months, I have been working under the mentorship of [Alban Crequy](https://github.com/alban) and [Burak Ok](https://github.com/burak-ok) as part of the [LFX Mentorship](https://mentorship.lfx.linuxfoundation.org/project/7fdda09c-0eb8-466b-9fdf-e4b3c6a1d5b3) program.
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+Consider this blog a report on my work during the mentorship.
+
+<!-- truncate -->
+
+## What have I been up to?
+
+My project involved developing a [new gadget](https://github.com/inspektor-gadget/inspektor-gadget/issues/3194) for detecting potential deadlocks in multi-threaded user space applications, using eBPF and WASM.
+
+Debugging and preventing deadlocks in modern, complex codebases can be challenging, and this gadget simplifies the process by identifying risks early, promoting safer and more efficient resource management.
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:20]: I think splitting this sentence into two sentences would make it easier for users to follow.

```suggestion
Debugging and preventing deadlocks in modern, complex codebases can be challenging. This gadget simplifies the process by identifying risks early, promoting safer and more efficient resource management.
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+## Who are Inspektors and what are they inspekting?
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:22]: Maybe even remove the title and move the paragraph to the first subchapter. Would love additional feedback on this
```suggestion
## What is Inspektor Gadget and what is getting inspekted?
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+If you’re new here, welcome! [Inspektor Gadget](https://www.inspektor-gadget.io/) is a systems inspection framework leveraging eBPF-powered gadgets encapsulated into deployable [OCI images](https://opencontainers.org/) to provide powerful observability and debugging capabilities across environments, from container runtimes to multi-node Kubernetes clusters, with everything you need to collect, filter, format and export valuable systems data. Check out some of these gadgets [here](https://www.inspektor-gadget.io/docs/latest/gadgets/). Moving on...
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:24]: Would also split this into multiple smaller sentences for readability

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+## What’s a deadlock?
+
+A [deadlock](https://en.wikipedia.org/wiki/Deadlock_(computer_science)) in a system occurs when a process becomes stuck, unable to proceed because each thread (sub-process) is waiting for a resource that another thread is holding. Since none of the threads can continue, the system remains in a state of permanent lock-up unless some external action is taken.
+
+Let’s take an example of a process with three mutexes and three threads.
+A mutex (_mutual exclusion_) is an object used to control access to a mutable resource (variables, files, etc.) and can be locked by a thread to indicate that the resource is currently in use by that thread. If any other thread wants access to that resource, all they can do is wait for the previous thread to unlock that mutex. This ensures data consistency as no two threads modify the resource simultaneously.
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:31]:
```suggestion
A mutex (used for _mutual exclusion_) is an object used to control access to a mutable resource (variables, files, etc.) and can be locked by a thread to indicate that the resource is currently in use. If any other thread wants access to that resource, all they need to do is wait for the mutex to get unlocked. This ensures data consistency as no two threads modify the resource simultaneously.
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+Thread 1, currently using mutex 1, requests access to lock mutex 2, but has to wait as mutex 2 is being used by thread 2. Now thread 2 requests access to mutex 3, but also has to wait as mutex 3 is being used by thread 3. Finally, thread 3 requests access to mutex 1, but has to wait as, in case you forgot, mutex 1 is being used by thread 1.
+We have come full circle ending up with the threads being stuck in an infinite wait cycle due to mutex lock order inversions, thus hanging the process.
+
+:::tip
+
+The diagram above represents a directed graph. Keep reading to learn how we use it to detect these deadlocks.
+
+:::
+
+You can simulate a deadlock in code using the following C++ program adapted from [BCC tools](https://github.com/iovisor/bcc/blob/master/tools/deadlock_example.txt#L185):
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:46]: It never deadlocked on me :D
```suggestion
You can simulate a potential deadlock in code using the following C++ program adapted from [BCC tools](https://github.com/iovisor/bcc/blob/master/tools/deadlock_example.txt#L185):
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    std::this_thread::sleep_for(std::chrono::seconds(5));
+    std::cout << "done!" << std::endl;
+}
+```
+
+## The Gadget
+
+Inspektor Gadget recently unlocked support for uprobe based gadgets, as part of another LFX Mentorship project by fellow inspektor [Tianyi Liu](https://github.com/i-Pear) in the spring term. Read more about it [here](https://www.inspektor-gadget.io/blog/2024/06/supporting-uprobe-based-gadgets-lfx-mentorship-report).
+
+Using uprobes, we attach to the `pthread_mutex_lock` and `pthread_mutex_unlock` methods from the `pthread` library ie. POSIX threads (now integrated into the GNU’s standard `libc` library), as follows:
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:102]:
```suggestion
Using uprobes, we attach to the `pthread_mutex_lock` and `pthread_mutex_unlock` functions from the `pthread` library ie. POSIX threads (now integrated into the GNU’s standard `libc` library), as follows:
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+```c
+SEC("uprobe/libc:pthread_mutex_lock")
+int BPF_UPROBE(trace_uprobe_mutex_lock, void *mutex_addr)
+{
+    // Intercept it as you want!
+}
+```
+
+[POSIX threads](https://man7.org/linux/man-pages/man7/pthreads.7.html) is the standard threading library on Unix-like systems, providing low-level control over thread creation, synchronization, and management.
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:112]:
```suggestion
[POSIX threads](https://man7.org/linux/man-pages/man7/pthreads.7.html) is the standard threading library on Unix-like systems, providing low-level control over thread creation, synchronization and management.
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    
+    std::cout << "sleeping to allow trace to collect data..." << std::endl;
+    std::this_thread::sleep_for(std::chrono::seconds(5));
+    std::cout << "done!" << std::endl;
+}
+```
+
+## The Gadget
+
+Inspektor Gadget recently unlocked support for uprobe based gadgets, as part of another LFX Mentorship project by fellow inspektor [Tianyi Liu](https://github.com/i-Pear) in the spring term. Read more about it [here](https://www.inspektor-gadget.io/blog/2024/06/supporting-uprobe-based-gadgets-lfx-mentorship-report).
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:100]:
```suggestion
Inspektor Gadget recently unlocked support for uprobe based gadgets, as part of another LFX Mentorship project by fellow Inspektor [Tianyi Liu](https://github.com/i-Pear) in the spring term. Read more about it [here](https://www.inspektor-gadget.io/blog/2024/06/supporting-uprobe-based-gadgets-lfx-mentorship-report).
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+SEC("uprobe/libc:pthread_mutex_lock")
+int BPF_UPROBE(trace_uprobe_mutex_lock, void *mutex_addr)
+{
+    // Intercept it as you want!
+}
+```
+
+[POSIX threads](https://man7.org/linux/man-pages/man7/pthreads.7.html) is the standard threading library on Unix-like systems, providing low-level control over thread creation, synchronization, and management.
+
+Once attached, it now allows us to intercept and trace mutex acquisitions and releases as they happen. We then build a mutex wait directed graph of all mutexes and threads, the same on we saw above in the diagram, and the deadlock can simply be detected by detecting cycles in the di-graph using algorithms like the [_Tarjan’s_](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm) and the [_Johnson’s_](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm).
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:114]: I think for users and readers the cycle detection algorithm used might not be that relevant
```suggestion
Once attached, it now allows us to intercept and trace mutex acquisitions and releases as they happen. We then build a mutex wait directed graph of all mutexes and threads, the same one we saw above. After that a potential deadlock can be simply detected by finding cycles in the directed graph.
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+Here’s a high level overview of how the gadget is detecting potential deadlocks:
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:116]:
```suggestion
Here’s a simple high level overview of how the gadget is detecting potential deadlocks:
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+![A flowchart depicting the deadlock detection process](/media/2024-12-06-deadlock-detection.svg)
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:118]:
```suggestion
![A flowchart depicting the deadlock detection process in the new gadget](/media/2024-12-06-deadlock-detection.svg)
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+This gadget detects "potential" deadlocks and not just actual deadlocks as a deadlock that didn't happen during the trace could happen in the future.
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:120]:
```suggestion
This gadget detects "potential" deadlocks and not just actual deadlocks, as a deadlock that didn't happen during the trace could happen in the future.
```

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+A deadlock occurring depends on various factors such as thread scheduling and order of mutex acquisitions/releases, and hence it makes sense to report mutex lock order inversions as potentially unsafe resource management.
```
Comment (burak-ok) [inline blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:121]: Great explanation!

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md (hunk 4 above)
Comment (Snehil-Shah) [reply blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:22]: The title is actually a casual phrase from an internet meme😬, typically used to talk about something which the reader has absolutely no idea about.

An example scenario: say you have never heard of Bluetooth before.. then you might ask:

"Who is Bluetooth, and why does everyone keep connecting to them?" 😅

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md (hunk 5 above)
Comment (Snehil-Shah) [reply blog/2024-12-06-deadlock-gadget-lfx-mentorship.md]: Done. Is it fine now?

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md (hunk 11 above)
Comment (Snehil-Shah) [reply blog/2024-12-06-deadlock-gadget-lfx-mentorship.md]: Done! Also phrased it a bit differently, if that's fine..

File: blog/2024-12-06-deadlock-gadget-lfx-mentorship.md (hunk 9 above)
Comment (burak-ok) [reply blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:118]: I have no idea :sweat_smile:

Comment (Snehil-Shah) [reply blog/2024-12-06-deadlock-gadget-lfx-mentorship.md:118]: Let's keep it then😅? Feels more readable to me..

**Review by burak-ok — APPROVED** [review]
LGTM. Thanks :)

**Pull request merged**